core/src/main/scala/org/apache/spark/deploy/history/ApplicationCache.scala

```diff
@@ -48,11 +49,28 @@ private[history] class ApplicationCache(
     val retainedApplications: Int,
     val clock: Clock) extends Logging {
 
+  /**
+   * Keep track of SparkUIs in [[ApplicationCache#appCache]] and SparkUIs removed from
+   * [[ApplicationCache#appCache]] but not detached yet.
+   */
+  private val loadedApps = new ConcurrentHashMap[CacheKey, CountDownLatch]()
+
   private val appLoader = new CacheLoader[CacheKey, CacheEntry] {
 
     /** the cache key doesn't match a cached entry, or the entry is out-of-date, so load it. */
     override def load(key: CacheKey): CacheEntry = {
-      loadApplicationEntry(key.appId, key.attemptId)
+      // Ensure old SparkUI has been detached before loading new one.
```

Comment:
I have some questions about this. Shouldn't the `CacheLoader#load` method only be executed once for accessing the same key? Why is there a scenario where the same key enters the load method more than once? 

Is this a bug of Guava 14.0.1 or did I have an incorrect understanding of the mechanism of `CacheLoader`? Please correct me.

 In addition, from the Jira description, does this issue only exist in Spark 3.1.2?

also cc @steveloughran FYI due to he is the author of `ApplicationCache`.

Reply:
@zhouyifan279 After reading the pr description again with the new UT, I understand your intention. Let me think more about it

Reply:
Yeah, already understand. Let me think more about this fix

Reply:
> @zhouyifan279 By the way, does RocksDB backend also have the issue you pointed out?

Active SparkUI's KVStore is closed when this issue occurs. So RocksDB backend also have the issue.